multicluster/service-mirror/cluster_watcher.go

```diff
@@ -514,6 +547,13 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateService(service *corev1.S
 			_, isMirroredRes := localSvc.Labels[consts.MirroredResourceLabel]
 			clusterName := localSvc.Labels[consts.RemoteClusterNameLabel]
 			if isMirroredRes && (clusterName == rcsw.link.TargetClusterName) {
+				// If headless, also delete its nested services.
+				if rcsw.headlessServicesEnabled && localSvc.Spec.ClusterIP == corev1.ClusterIPNone {
```

Comment:
Yeah, I think it would make sense and it'd work well. I considered it too at first, but I had some reservations.

First, I like that in `createOrUpdateService` we already fetch the Mirror Service Spec to check whether it should be cleaned up. This makes it pretty inconsequential to check if the service is also headless. By contrast, in `handleRemoteServiceDeleted` we don't readily have access to the service's spec; for each Mirror Service that we delete, we would have to also fetch its object to see whether it is headless. 

Second, it felt more intuitive and correct to me to delete the Endpoint Mirrors first before we delete the Headless Mirror itself. This came from how retries are processed. In `handleRemoteServiceDeleted` we would introduce another point of failure when creating events to delete the Endpoint Mirrors (done in `deleteHeadlessMirrors`). 
   * If we first delete the Headless Mirror then we cannot delete the Endpoint Mirrors in the same way (since we will no longer have access to the service's Endpoints object). 
   * If we first delete the Endpoint Mirrors and then we fail to delete the Headless Mirror, then we will retry to delete the Headless Mirror which will consequently add duplicate events for the Endpoint Mirrors in the queue. At least this is how it looks like to me. 

Third, I thought it would complicate `handleRemoteServiceDeleted` a bit too much (and give it more than one thing to do: delete services but also check if they're headless). We already do this headless check in `handleOnDelete`; because of this, it felt like this headless check should be done for all "entry points" to `handleRemoteServiceDeleted` (which would be `handleOnDelete` and `createOrUpdateService`). 

---

Typing this out made me think of a different approach though. When I changed the diffing logic, I added a `root-headless` label to all Endpoint Mirrors -- this label is meant to group them under the same Headless Mirror (gives us a way to identify which headless service they belong to). I think a better approach to deletion would be to use a label selector to get all Endpoint Mirrors that correspond to the same Headless Mirror and create events based on the list of services we get back. Let me exemplify:

In `handleRemoteServiceDeleted`:
```
// retrieve service spec to see if it's headless
service, err := CoreV1().Services(ev.Namespace).Get(ev.name)
// err handling redacted

var endpointMirrorss []corev1.Services
if service.Spec.ClusterIP == clusterIPNone {
  endpointMirrors, err := rcsw.localAPIClient.Svc().Lister().List(labels.Set(matchLabels).AsSelector())
}

// delete service like before
err = Corev1().(...)
// any errors? retry

for _, svc := range endpointMirrors {
     rcsw.eventsQueue.Add( // ... )
}
```

If we keep the logic around this in `handleRemoteServiceDeleted` then we can get rid of `deleteHeadlessMirrors` entirely, get rid of the logic around Headless Mirrors in `handleOnDelete` and also remove the logic from `createOrUpdateService` as suggested.

@adleong what do you think? Would this make sense, and do my concerns make since in this context? Sorry in advance for the long post :)